File: src/models/technologies.ts
import { TechStack, LanguageInfo, FileNode } from '@/types'
import { getTechDatabase } from './knowledge'

interface DetectionResult {
  name: string
  category: string
  confidence: number
  evidence: string[]
}

export function detectTechnologies(
  languages: Record<string, number>,
  fileTree: FileNode[],
  dependencyFiles: Record<string, string>,
  readmeContent: string,
  topics: string[]
): TechStack {
  const detections: DetectionResult[] = []
  const allFiles = collectFiles(fileTree)

  for (const tech of getTechDatabase()) {
    const evidence: string[] = []
    let confidence = 0

    for (const pattern of tech.patterns) {
      try {
        const regex = new RegExp(pattern, 'i')

        for (const file of allFiles) {
          if (regex.test(file)) {
            evidence.push(file)
            break
          }
        }

        for (const [filepath, content] of Object.entries(dependencyFiles)) {
          if (regex.test(filepath) || regex.test(content)) {
            evidence.push(filepath)
            break
          }
        }
      } catch {
      }
    }

    if (evidence.length > 0) {
      confidence = Math.min(tech.confidence, 50 + evidence.length * 15)
      detections.push({ name: tech.name, category: tech.category, confidence, evidence })
    }
  }

  if (readmeContent) {
    const readmeLower = readmeContent.toLowerCase()
    for (const tech of getTechDatabase()) {
      const existing = detections.find(d => d.name === tech.name)
      if (existing) {
        for (const pattern of tech.patterns) {
          try {
            const regex = new RegExp(pattern.replace(/\\/g, ''), 'i')
            if (regex.test(readmeLower)) {
              existing.confidence = Math.min(100, existing.confidence + 5)
            }
          } catch {
          }
        }
      } else {
        const hasMatch = tech.patterns.some(p => {
          try {
            const cleanPattern = p.replace(/\\\./g, '.').replace(/\\$/g, '')
            return readmeLower.includes(cleanPattern.toLowerCase())
          } catch {
            return false
          }
        })
        if (hasMatch) {
          detections.push({ name: tech.name, category: tech.category, confidence: 60, evidence: ['README mention'] })
        }
      }
    }
  }

  for (const topic of topics) {
    const matched = getTechDatabase().find(t =>
      t.name.toLowerCase() === topic.toLowerCase()
    )
    if (matched && !detections.find(d => d.name === matched.name)) {
      detections.push({ name: matched.name, category: matched.category, confidence: 70, evidence: ['GitHub topic'] })
    }
  }

  const languageEntries: LanguageInfo[] = Object.entries(languages)
    .sort(([, a], [, b]) => b - a)
    .map(([name, bytes]) => ({
      name,
      percentage: Math.round((bytes / Math.max(1, Object.values(languages).reduce((a, b) => a + b, 0))) * 1000) / 10,
      bytes,
    }))

  const techStack: TechStack = {
    languages: languageEntries,
    frameworks: [],
    databases: [],
    tools: [],
    infrastructure: [],
  }

  const grouped: Record<string, DetectionResult[]> = {}
  for (const d of detections) {
    if (!grouped[d.category]) grouped[d.category] = []
    grouped[d.category].push(d)
  }

  for (const [category, items] of Object.entries(grouped)) {
    const sorted = items.sort((a, b) => b.confidence - a.confidence).slice(0, 8)

    if (category === 'language') {
      for (const item of sorted) {
        if (!techStack.languages.find(l => l.name === item.name)) {
          techStack.languages.push({
            name: item.name,
            percentage: 0,
            bytes: 0,
          })
        }
      }
    } else if (category === 'framework') {
      techStack.frameworks.push(...sorted.map(s => `${s.name} (${s.confidence}%)`))
    } else if (category === 'database') {
      techStack.databases.push(...sorted.map(s => `${s.name} (${s.confidence}%)`))
    } else if (category === 'tool') {
      techStack.tools.push(...sorted.map(s => `${s.name} (${s.confidence}%)`))
    } else if (category === 'infrastructure') {
      techStack.infrastructure.push(...sorted.map(s => `${s.name} (${s.confidence}%)`))
    }
  }

  return techStack
}

function collectFiles(tree: FileNode[], path: string = ''): string[] {
  const files: string[] = []
  const MAX_FILES = 5000
  for (const node of tree) {
    if (files.length >= MAX_FILES) break
    const fullPath = path ? `${path}/${node.name}` : node.name
    if (node.type === 'blob') {
      files.push(fullPath)
    }
    if (node.children && files.length < MAX_FILES) {
      const childFiles = collectFiles(node.children, fullPath)
      for (const f of childFiles) {
        if (files.length >= MAX_FILES) break
        files.push(f)
      }
    }
  }
  return files
}

export function detectPackageManager(dependencyFiles: Record<string, string>): string {
  const files = Object.keys(dependencyFiles)
  if (files.some(f => f.includes('package-lock'))) return 'npm'
  if (files.some(f => f.includes('yarn.lock'))) return 'yarn'
  if (files.some(f => f.includes('pnpm-lock'))) return 'pnpm'
  if (files.some(f => f.includes('Cargo'))) return 'cargo'
  if (files.some(f => f.includes('Gemfile'))) return 'bundler'
  if (files.some(f => f.includes('requirements'))) return 'pip'
  if (files.some(f => f.includes('go.mod'))) return 'go mod'
  if (files.some(f => f.includes('pom.xml'))) return 'maven'
  if (files.some(f => f.includes('build.gradle'))) return 'gradle'
  return 'npm'
}

export function detectDevCommands(dependencyFiles: Record<string, string>, languages: Record<string, number>): {
  dev: string
  build: string
  test: string
  install: string
} {
  const hasNode = languages.hasOwnProperty('JavaScript') || languages.hasOwnProperty('TypeScript')
  const hasPython = languages.hasOwnProperty('Python')
  const hasGo = languages.hasOwnProperty('Go')
  const hasRust = languages.hasOwnProperty('Rust')
  const depFiles = Object.keys(dependencyFiles)

  if (depFiles.some(f => f.includes('package.json'))) {
    return { dev: 'npm run dev', build: 'npm run build', test: 'npm test', install: 'npm install' }
  }
  if (hasPython || depFiles.some(f => f.includes('requirements') || f.includes('Pipfile') || f.includes('pyproject'))) {
    return { dev: 'python main.py', build: 'python setup.py build', test: 'pytest', install: 'pip install -r requirements.txt' }
  }
  if (hasGo || depFiles.some(f => f.includes('go.mod'))) {
    return { dev: 'go run .', build: 'go build ./...', test: 'go test ./...', install: 'go mod download' }
  }
  if (hasRust || depFiles.some(f => f.includes('Cargo.toml'))) {
    return { dev: 'cargo run', build: 'cargo build --release', test: 'cargo test', install: 'cargo build' }
  }
  if (hasNode) {
    return { dev: 'npm run dev', build: 'npm run build', test: 'npm test', install: 'npm install' }
  }
  return { dev: 'npm run dev', build: 'npm run build', test: 'npm test', install: 'npm install' }
}

File: backend/models/technologies.py
from __future__ import annotations
import re
from .knowledge import get_tech_database


def collect_files(tree: list[dict], prefix: str = '') -> list[str]:
    files: list[str] = []
    MAX_FILES = 5000
    for node in tree:
        if len(files) >= MAX_FILES:
            break
        full = f"{prefix}/{node['name']}" if prefix else node['name']
        if node.get('type') == 'blob':
            files.append(full)
        if node.get('children') and len(files) < MAX_FILES:
            files.extend(collect_files(node['children'], full))
    return files


def detect_technologies(
    languages: dict[str, int],
    file_tree: list[dict],
    dependency_files: dict[str, str],
    readme_content: str,
    topics: list[str],
) -> dict:
    detections: list[dict] = []
    all_files = collect_files(file_tree)

    for tech in get_tech_database():
        evidence: list[str] = []
        for pattern in tech['patterns']:
            try:
                regex = re.compile(pattern, re.IGNORECASE)
                for f in all_files:
                    if regex.search(f):
                        evidence.append(f)
                        break
                for fp, content in dependency_files.items():
                    if regex.search(fp) or regex.search(content):
                        evidence.append(fp)
                        break
            except re.error:
                pass
        if evidence:
            conf = min(tech['confidence'], 50 + len(evidence) * 15)
            detections.append({'name': tech['name'], 'category': tech['category'],
                               'confidence': conf, 'evidence': evidence})

    if readme_content:
        readme_lower = readme_content.lower()
        for tech in get_tech_database():
            existing = next((d for d in detections if d['name'] == tech['name']), None)
            if existing:
                for pattern in tech['patterns']:
                    try:
                        clean = pattern.replace('\\', '')
                        if re.search(clean, readme_lower, re.IGNORECASE):
                            existing['confidence'] = min(100, existing['confidence'] + 5)
                    except re.error:
                        pass
            else:
                for pattern in tech['patterns']:
                    try:
                        clean = pattern.replace('\\.', '.').replace('\\', '')
                        if clean.lower() in readme_lower:
                            detections.append({'name': tech['name'], 'category': tech['category'],
                                               'confidence': 60, 'evidence': ['README mention']})
                            break
                    except re.error:
                        pass

    for topic in topics:
        matched = next((t for t in get_tech_database()
                       if t['name'].lower() == topic.lower()), None)
        if matched and not any(d['name'] == matched['name'] for d in detections):
            detections.append({'name': matched['name'], 'category': matched['category'],
                               'confidence': 70, 'evidence': ['GitHub topic']})

    total_bytes = sum(languages.values()) or 1
    lang_entries = sorted(
        [{'name': k, 'percentage': round((v / total_bytes) * 1000) / 10, 'bytes': v}
         for k, v in languages.items()],
        key=lambda x: -x['bytes']
    )

    result = {'languages': lang_entries, 'frameworks': [], 'databases': [],
              'tools': [], 'infrastructure': []}
    grouped: dict[str, list[dict]] = {}
    for d in detections:
        grouped.setdefault(d['category'], []).append(d)
    for cat, items in grouped.items():
        sorted_items = sorted(items, key=lambda x: -x['confidence'])[:8]
        if cat == 'language':
            for item in sorted_items:
                if not any(l['name'] == item['name'] for l in result['languages']):
                    result['languages'].append({'name': item['name'], 'percentage': 0, 'bytes': 0})
        elif cat in result:
            result[cat].extend(f"{s['name']} ({s['confidence']}%)" for s in sorted_items)
    return result


def detect_package_manager(dependency_files: dict[str, str]) -> str:
    files = list(dependency_files.keys())
    if any('package-lock' in f for f in files): return 'npm'
    if any('yarn.lock' in f for f in files): return 'yarn'
    if any('pnpm-lock' in f for f in files): return 'pnpm'
    if any('Cargo' in f for f in files): return 'cargo'
    if any('Gemfile' in f for f in files): return 'bundler'
    if any('requirements' in f for f in files): return 'pip'
    if any('go.mod' in f for f in files): return 'go mod'
    if any('pom.xml' in f for f in files): return 'maven'
    if any('build.gradle' in f for f in files): return 'gradle'
    return 'npm'


def detect_dev_commands(dependency_files: dict[str, str],
                        languages: dict[str, int]) -> dict[str, str]:
    has_node = 'JavaScript' in languages or 'TypeScript' in languages
    has_python = 'Python' in languages
    has_go = 'Go' in languages
    has_rust = 'Rust' in languages
    dep_files = list(dependency_files.keys())
    if any('package.json' in f for f in dep_files):
        return {'dev': 'npm run dev', 'build': 'npm run build',
                'test': 'npm test', 'install': 'npm install'}
    if has_python or any(k in ' '.join(dep_files) for k in ['requirements', 'Pipfile', 'pyproject']):
        return {'dev': 'python main.py', 'build': 'python setup.py build',
                'test': 'pytest', 'install': 'pip install -r requirements.txt'}
    if has_go or any('go.mod' in f for f in dep_files):
        return {'dev': 'go run .', 'build': 'go build ./...',
                'test': 'go test ./...', 'install': 'go mod download'}
    if has_rust or any('Cargo.toml' in f for f in dep_files):
        return {'dev': 'cargo run', 'build': 'cargo build --release',
                'test': 'cargo test', 'install': 'cargo build'}
    if has_node:
        return {'dev': 'npm run dev', 'build': 'npm run build',
                'test': 'npm test', 'install': 'npm install'}
    return {'dev': 'npm run dev', 'build': 'npm run build',
            'test': 'npm test', 'install': 'npm install'}
